Added vectors (straight from sspas)

Grissess opened the pull request:
The usual modus operandi is to include a vector struct (not a pointer thereof) in whatever you want (id est, type composition), then pass the pointer to that member to the functions.

Not tested, but looks pretty clean.

Grissess committed:
Added vectors (straight from sspas)

sjrct commented:
Seems good. I am guessing its usage in sspas has served as a test case, but probably should write a test file to demonstrate usage and prevent future errors from modification, though it does seem pretty easy to use.

Merge if you so desire.

Grissess commented:
Grissess/sspas@81d631aa3d5ee28eabcd2d689dd29ebc1f0f46c7 doesn't build right now--that's something I should definitely fix within the coming weeks, as it is an academic project :P 

The comment above refers to a line in the file that grows the buffer; right now, I do it in powers of two, which can be a little inefficient (I seem to remember someone recommending the golden ratio). In Grissess/dsl@10976abb77cee1f9beb6514a80e38e950bca9a0d, I'm dealing with that by providing the user a `dsl_overallocation` (see https://github.com/Grissess/dsl/blob/master/array.c#L110). Despite that, Sol (the only real consumer of this library) uses only the default settings :P .

I digress, that's an issue. Usage looks like some of the lines in sspas, linked above (see particularly https://github.com/Grissess/sspas/blob/master/type.c):

```
typedef struct { ... } some_object;
typedef struct {
    vector keys; /* Of void * */
    vector values; /* Of some_object * */
} silly_assoc_arr;

void sar_init(silly_assoc_arr *sar) {
    vec_init(&sar->keys);
    vec_init(&sar->values);
}

some_object *sar_get(silly_assoc_arr *sar, void *key) {
    /* There's a lot of useful functions like this in here */
    ssize_t idx = vec_search(&sar->keys, key);
    if(idx < 0) return NULL;
    /* No vec_get/vec_set yet, that's overhead--unless you inline it, I s'pose :) */
    return ((some_object **) sar->values.buf)[idx]; /* Note the prudent cast */
}
```

That get syntax is a little disgusting for my tastes; I might macro it if no one else minds :P

Grissess committed:
Added vector getters/setters
(also fixed .gitignore for Vim and added variadic macro helpers)
(...it does compile, by the way :P)

Grissess commented:
...with the above commit, you could define `sar_get` as simply:

```
some_object *sar_get(silly_assoc_arr *sar, void *key) {
    ssize_t idx = vec_search(&sar->keys, key);
    if(idx < 0) return NULL;
    return vec_get(&sar->values, idx, some_object);
}
```

Due to some macro magic, if you have a vector of void pointers, you can elide the third argument :P

Grissess committed:
Whoops, let's use the right struct op

Grissess committed:
Bug fixes galore, and a vector test suite

Grissess commented:
OK, I'd say it's tested now--it passes the rigorous test now in test/vector/ :P

Grissess commented:
One last thing before I head elsewhere (I promise!): I notice you have the ability to pass userdata in your enhanced list API to the iter function; I have that prototyped out as type `vec_iter_f`, which only accepts the element. One _can_ abuse `vec_test` and `vec_test_f` (which receives userdata, the element, and the vector and index into it) into doing something similar--with the important caveat that test function return nonzero values to bubble, so such a "test iterator" would have to return 0. It seems a little dirty...

Should I change that prototype? It would be trivial, but it would mean that I couldn't implement the test `free_data` as `vec_foreach(v, free)` :P

sjrct commented:
I do not think changing the prototype would be a bad thing, it essentially allows one to use the functionality of a fold function and so forth. I would prefer to avoid the abusing `vec_test*` in such a manner as it would indeed be a little dirty. I needed something like that in the server implementation stuff (the userdata), so there are certainly use cases, but it overall it does not seem like a high priority change.

Grissess committed:
Added second argument to vec_map

Grissess commented:
It is done.

Grissess committed:
Also added new arg to vec_foreach

sjrct commented:
This seems pretty good, feel free to merge.
